solutions/gold/csa-MountainTime.mdx

```diff
@@ -0,0 +1,160 @@
+---
+id: csa-MountainTime
+source: CSA
+title: Mountain Time
+author: Ryan Chou
+---
+
+<Spoiler title="Hint 1">
+
+Instead of thinking of the problem as a grid, think of it as a graph, where we're adding one cell at a time.
+
+In what order should we process these cells?
+
+</Spoiler>
+
+<Spoiler title="Answer to Hint 1">
+
+We'll process them in decreasing order of height.
+
+</Spoiler>
+
+<Spoiler title="Hint 2">
+
+Whenever we process a cell, we'll want to merge it with all of its neighbors that have already been added (taller than it).
+
+What data structure should we use such that we can efficiently merge two components?
+
+</Spoiler>
+
+<Spoiler title="Answer to Hint 2">
+
+We can use a [DSU](/gold/dsu) to solve the problem.
+
+</Spoiler>
+
+<Spoiler title="Solution">
+
+## Explanation
+
+Let's imagine that we're currently at cell $x$, which is part of component $A$, and we're merging with some adjacent cell $y$, which is part of component $B$.
+
+We can break it into two cases.
+
+1. The tallest mountains in $A$ and $B$ are the **same** height. In this case, we'll combine all the cells into one component. In the implementation, we use small-to-large merging to minimize the number of cells we have to update.
+
+2. The tallest mountains in $A$ and $B$ are of **different** heights. Without loss of generality, let's assume that mountain $A$ has the tallest mountain. Then, we know the answer for all the cells in component $B$ because $x$ is the tallest mountain that can connect them both.
+
+To simplify the process, we'll root each component at its maximal cell and store indices of components directly in the DSU. This way, we can easily iterate over cells
+we've found the answer for.
```

Comment:
```suggestion
Let's imagine we're currently at cell $x$, part of component $A$, and merging with some adjacent cell $y$, part of component $B$.	 

We can break it into two cases.	 

1. The tallest mountains in $A$ and $B$ are the **same** height. In this case, we'll combine all the cells into one component using small-to-large merging to minimize the number of cells we must update.	 
  
2. The tallest mountains in $A$ and $B$ are of **different** heights. Without loss of generality, let's assume that mountain $A$ has the tallest mountain; we know the answer for the cells in component $B$ because $x$ is the tallest mountain that can connect them both.	 
  
To simplify the process, we'll root each component at its maximal cell and store their corresponding indices in the DSU. Thus, we can easily iterate over the cells for which we've found the answer.
```